Add a short extra lesson on removing files/datasets (#35)
psychoinformatics-de/rdm-course: _episodes/92-filesystem-operations.md

~~~
# make sure to run this command outside of any existing dataset
datalad clone https://github.com/datalad-datasets/machinelearning-books.git
~~~
{: .language-bash}

The PDFs in this dataset are all annexed contents, and can be retrieved using ``datalad get``:

~~~
cd machinelearning-books
datalad get A.Shashua-Introduction_to_Machine_Learning.pdf
~~~
{: .language-bash}

Once you have obtained one file's contents, they are known to exist in their remote registered location and locally on your system:

~~~
git annex whereis A.Shashua-Introduction_to_Machine_Learning.pdf
~~~
{: .language-bash}

~~~
whereis A.Shashua-Introduction_to_Machine_Learning.pdf (2 copies) 
  	00000000-0000-0000-0000-000000000001 -- web
   	0d757ca9-ea20-4646-96cb-19dccd732f8c -- adina@muninn:/tmp/machinelearning-books [here]

  web: https://arxiv.org/pdf/0904.3664v1.pdf
ok
~~~
{: .output}

This output shows that the file contents of this book exist locally and "on the web", more precisely under the URL [arxiv.org/pdf/0904.3664v1.pdf](https://arxiv.org/pdf/0904.3664v1.pdf).

A ``datalad drop <file>`` performs an internal check if the to-be-dropped file content could be re-obtained before dropping it.

> When the file content exists in more than your local system, _and_ at least one of those other location is successfully probed to have this content, dropping its file content succeeds.

Reviewer: more places/locations than?

Author: oops, missed this one, thanks!

Author: pushed a fix

Here is a successful ``drop``:

~~~
datalad drop A.Shashua-Introduction_to_Machine_Learning.pdf
~~~
{: .language-bash}

~~~
drop(ok): A.Shashua-Introduction_to_Machine_Learning.pdf (file)
~~~
{: .output}

### ``remove`` the easy way: datasets with remote availability

> To demonstrate removing a datasets, let's remove the `machinelearning-books` dataset.

Reviewer: ```suggestion
To demonstrate removing datasets, let's remove the `machinelearning-books` dataset.
```

The ``--reckless`` parameter has multiple modes. Each mode is one type of safety check. Because it is the "availability" check fails, ``--reckless availability`` will disable it:

~~~
datalad drop local-file.txt --reckless availability
~~~
{: .language-bash}

The file content of ``local-file.txt`` is irretrievably gone now.

> ## Dropping multiple versions of a file
> If a file has been updated and saved multiple times, the revision history of the dataset will have each version in the annex, and only the most recent versions' content would be dropped.
>Consider the following example:
> {% raw %}
>       datalad run -m "Make a first version" --output local-file-multiple-revision.txt "echo 'This is the first revision' > {outputs}"
>       # overwrite the contents of the file 
>       datalad run -m "Make a second version" --output local-file-multiple-revision.txt "echo 'This is the second revision' > {outputs}"
> {% endraw %}
>

> >``datalad drop --reckless availability local-file-multiple-revision.txt`` would drop the _current_ revision of the file, but the first version (with the content "``This is the first revision``") would still exist in the datasets history.

Reviewer: ```suggestion
>``datalad drop --reckless availability local-file-multiple-revision.txt`` would drop the _current_ revision of the file, but the first version (with the content "``This is the first revision``") would still exist in the dataset's history.
```

Author: thanks for catching this!

Removing datasets without ``datalad remove`` is a bad idea - for one, because it skips safety checks that aim to prevent data loss, and secondly, because it doesn't work so well.
If you were to remove a dataset using common Unix commands such as ``rm -rf`` or your operating systems' file browser, you would see the residue of write-protected annexed files refusing to be deleted.
On the command line, it would look something like this:

~~~
cannot remove 'machinelearning-books/.git/annex/objects/zw/WF/URL-s1787416--http&c%%www1.maths.leeds.ac.uk%,126charles%statlog%whole.pdf/URL-s1787416--http&c%%www1.maths.leeds.ac.uk%,126charles%statlog%whole.pdf': Permission denied
rm: cannot remove 'machinelearning-books/.git/annex/objects/jQ/GM/MD5E-s21322662--8689c3c26c3a1ceb60c1ba995d638677.pdf/MD5E-s21322662--8689c3c26c3a1ceb60c1ba995d638677.pdf': Permission denied
[...]
~~~
{: .output}

Afterwards, a few left-over but unusable dataset ruins remain.
To remove those from your system, you need to make the remaining files writable:

~~~
# add write permissions to the directory
chmod -R +w <dataset>
# remove everything that is left
rm -rf <dataset>
~~~
{: .language-bash}
## Further reading

> This overview only covers the surface of all possible file system operations, and not even all ways in which files or datasets can be removed. For a more complete overview, checkout [handbook.datalad.org/basics/101-136-filesystem.html](http://handbook.datalad.org/en/latest/basics/101-136-filesystem.html) - the navigation bar on the left of the page lists a number of useful file system operations.

Reviewer: ```suggestion
This overview only covers the surface of all possible file system operations, and not even all ways in which files or datasets can be removed. For a more complete overview, check out [handbook.datalad.org/basics/101-136-filesystem.html](http://handbook.datalad.org/en/latest/basics/101-136-filesystem.html) - the navigation bar on the left of the page lists a number of useful file system operations.
```
